### DISCUSSION TITLE
Tailwind classes component composition

### DISCUSSION THREAD

**Discussion by tanishqkancharla** [tailwindlabs/tailwindcss#3908] [Ideas]
Hello! I'm new to tailwind but to me, the CSS bloat in my JSX is a real problem. I hate seeing potentially dozens of classes in my components. I deal with it by simply moving out the classes into css files and defining my own classes:

```css
.block-link-lg {
  @apply block p-2 my-3 no-underline rounded border border-gray-300;
  @apply transition duration-200 ease-in-out hover:border-teal;
  @apply dark:border-gray-300-dark dark:hover:border-teal-dark;
}
```

I figured it would be a cool opportunity here for Tailwind to support this feature by basically letting me define compositions of styles directly in my config:

```json
{
  ...
  "block-link-lg": "block p-2 my-3 no-underline ..."
  }
```

It's not that different but you could take it a bit further by allowing even higher level compositions:

```json
{
  ...
  "block-clickable": "transition duration-200 ease-in-out border border-gray-300 hover:border-teal ..."
  "spaced-medium":  "p-2 my-3 ..."
  "block-link-lg": "block-clickable spaced-medium"
}
```

And then! Taking it *even* further, designating public/private styles to enforce a tight design system:

```json
{
  ...
  "private": [
    "block-clickable": "transition duration-200 ease-in-out border border-gray-300 hover:border-teal ...",
    "spaced-medium":  "p-2 my-3 ..."
  ],
  "public": [
    "block-link-lg": "block-clickable spaced-medium"
  ]
}
```

Again, I'm new to tailwind, so what I'm looking for may entirely be supported already somehow or in a plugin somewhere. But thought I would ask here and share my idea just in case.

**Comment by nanorepublica** [comment]
I mentioned this on the Tailwind Discord, but figured I would link to it here for completeness. 

I have made this plugin which addresses most if not all of these use cases: https://www.npmjs.com/package/tailwind-component-classes
Very happy for PRs and constructive feedback.
